osutil/disks/disks.go

```diff
@@ -0,0 +1,293 @@
+		}
+		m[strs[0]] = strs[1]
+	}
+
+	return m, scanner.Err()
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point. It
+// does a best effort in identifying partitions for the disk, but may be racy
+// due to races inherent in the initramfs surrounding udev and sysfs.
+func DiskFromMountPoint(mountpoint string, opts *Options) (Disk, error) {
+	// call the unexported version that may be mocked by tests
+	return diskFromMountPoint(mountpoint, opts)
+}
+
+// diskFromMountPointImpl uses the mount table, sysfs and udev to try and
+// identify the disk and full set of partitions for the specified mount point.
+// since during the initramfs things may be racy around udev adding new devices
+// this is a best effort search and partially matching partitions which do not
+// have the full information are discarded from the returned disk.
+func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	d := &disk{}
+	mountpointPart := partition{}
+	for _, mount := range mounts {
+		if mount.MountDir == mountpoint {
+			mountpointPart.major = mount.DevMajor
+			mountpointPart.minor = mount.DevMinor
+			mountpointPart.path = mount.MountSource
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
+	}
+
+	// TODO:UC20: if the mountpoint is of a decrypted mapper device, then we
+	//            need to trace back from the decrypted mapper device through
+	//            luks to find the real encrypted partition underneath the
+	//            decrypted one and thus the disk device for that partition
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(mountpointPart.path)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", mountpointPart.path, err)
+	}
+
+	// ID_PART_ENTRY_DISK will give us the major and minor of the disk that this
+	// partition originated from
+	if majorMinor, ok := props["ID_PART_ENTRY_DISK"]; ok {
+		maj, min, err := parseDeviceMajorMinor(majorMinor)
+		if err != nil {
+			// bad udev output?
+			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", mountpointPart.path, err)
+		}
+		d.major = maj
+		d.minor = min
+	} else {
+		// didn't find the property we need
+		return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", mountpointPart.path)
+	}
+
+	return d, nil
+
+}
+
+func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
+	// if we haven't found the partitions for this disk yet, do that now
+	if d.partitions == nil {
```

Comment:
it seems partitions could just be  mapping between  label -> uuid now, to preserve the current logic we wouldn't override a value in it once it is set first

Reply:
if we are going to make this more generalized for i.e. the DiskFromMountPoint function supporting volumes that are not partitions, then perhaps we should instead have a function `Partitions() ([]Partition,error)` for the `Disk` interface, and then the `Partition` interface has functions for `Label`, `Dev` (major/minor) and `PartitionUUID`, etc. ?

Reply:
given our current goal that's a bit overkill.  We can revisit later. I would keep it things as is for now, except DiskFromMountPoint should work even if the mounted thing is a disk and not a partition.